design/one-pager-fn-claim-conditions.md

```diff
@@ -0,0 +1,170 @@
+# Communication Between Composition Functions and the Claim
+
+* Owner: Dalton Hill (@dalton-hill-0)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
+
+## Background
+
+### Desired Behavior
+Composition Function authors should be able to communicate with users. 
+Topics of communication include:
+- The status of underlying resources.
+- Errors that need to be resolved by the user.
+- Internal errors.
+
+### Existing Behavior
+Currently the only path to communicate with the user is by defining a custom field in the Claim's status.
+For example, we can define an XRD with:
+```yaml
+status:
+  someCommunicationField:
+    - msg: "Something went wrong."
+```
+
+There are a couple issues with this solution.
+- If we need to halt resource reconciliation due to a fatal error, we can do so with the [SDK](https://github.com/crossplane/function-sdk-go)'s `response.Fatal`, however, this does not also allow us to update the XR and Claim for communication with the user.
+- There is an existing field that would be more intuitive to use as it is already performing this same task for Crossplane itself (`status.conditions`).
+
+## Proposal
+Allow the Composition Function author to set conditions inside the Claim's `status.conditions` field.
+
+From the Function author's perspective, they would just need to update the desired XR as follows:
+```go
+  // includes:
+  //    corev1 "k8s.io/api/core/v1"
+  //    metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+  //    xpv1   "github.com/crossplane/crossplane-runtime/apis/common/v1"
+  // 
+  //    "github.com/crossplane/function-sdk-go/response"
+  desiredXR, err := request.GetDesiredCompositeResource(req)
+  c := xpv1.Condition{
+      Type:               xpv1.ConditionType("ImageReady"),
+      Status:             corev1.ConditionFalse,
+      LastTransitionTime: metav1.Now(),
+      Reason:             "NotFound",
+      Message:            "The image provided does not exist or you are not authorized to use it.",
+  }
+  desiredXR.Resource.SetConditions(c)
+  response.SetDesiredCompositeResource(rsp, desiredXR)
+```
+
+Technically the behavior above is currently supported, though it has a couple limitations in its current form.
+- it only updates the XR's conditions (not the Claim)
+- it only updates if there were no fatal results returned by the function
```

Comment:
Did you check that this was the case? I know we thought it was, but I hadn't verified myself.

Reply:
Yes I verified this is the case (at least in the `master` branch).